main-command/src/main/scala/sbt/internal/classpath/ClassLoaderCache.scala

```diff
@@ -0,0 +1,217 @@
+package sbt.internal.classpath
+
+import java.io.File
+import java.lang.management.ManagementFactory
+import java.lang.ref.{ Reference, ReferenceQueue, SoftReference }
+import java.net.URLClassLoader
+import java.util.concurrent.atomic.AtomicInteger
+
+import sbt.internal.inc.classpath.{
+  AbstractClassLoaderCache,
+  ClassLoaderCache => IncClassLoaderCache
+}
+import sbt.internal.inc.{ AnalyzingCompiler, ZincUtil }
+import sbt.io.IO
+import xsbti.compile.{ ClasspathOptions, ScalaInstance }
+import xsbti.{ AppConfiguration, ExtendedScalaProvider }
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+private object ClassLoaderCache {
+  private def threadID = new AtomicInteger(0)
+}
+private[sbt] class ClassLoaderCache(val configuration: AppConfiguration)
```

Comment:
It might be good to add some unit tests for this new cache to test some of the invariants?

Reply:
Oh, I actually have a bunch of tests of the old ClassLoaderCache in the main project that I forgot to delete and won't compile now anyway. I'll update the PR tomorrow with the api changes needed to make them run in the command project.

Reply:
https://github.com/sbt/sbt/pull/4729/commits/53d609669cef1a1b0af1e4eb94b267a5d3538003 ports the old ClassLoaderCacheTest. The new cache doesn't support a size limit (yet) so I removed one of the tests. It isn't clear to me that there is much benefit to manually specifying the size limit since the actual size impact of each classloader can vary greatly and we can't easily measure it. For now, I think just having the jvm manage the cache implicitly through the soft references seems like the best approach until we are sure that it's a problem.